### PR TITLE
cls/journal: skip disconnected clients when calculating min_commit_position

### PR DESCRIPTION
**Pull request by trociny**
Fixes: https://tracker.ceph.com/issues/53888
Signed-off-by: Mykola Golub <mgolub@suse.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by MrFreezeex — APPROVED** [review]
I tested with my multipeer PR (https://github.com/ceph/ceph/pull/43359) and it works (both with 2 and 3 clusters), thanks!

File: src/cls/journal/cls_journal.cc [hunk 1]
```diff
@@ -271,11 +271,12 @@ int find_min_commit_position(cls_method_context_t hctx,
     }
 
     start_after = batch.rbegin()->id;
-
     // update the (minimum) commit position from this batch of clients
-    for(std::set<cls::journal::Client>::iterator it = batch.begin();
-        it != batch.end(); ++it) {
-      cls::journal::ObjectSetPosition object_set_position = (*it).commit_position;
+    for (auto &client : batch) {
```
Comment (idryomov) [inline src/cls/journal/cls_journal.cc:275]:
```suggestion
    for (const auto &client : batch) {
```

File: src/cls/journal/cls_journal.cc [hunk 2]
```diff
@@ -271,11 +271,12 @@ int find_min_commit_position(cls_method_context_t hctx,
...
+      if (client.state == cls::journal::CLIENT_STATE_DISCONNECTED) {
+        continue;
+      }
+      auto &object_set_position = client.commit_position;
```
Comment (idryomov) [inline src/cls/journal/cls_journal.cc:279]:
```suggestion
      const auto &object_set_position = client.commit_position;
```

**Review by idryomov — COMMENTED** [review]
Otherwise LGTM, although I'd prefer if the nice explanation from the tracker ticket was added to the commit message.

**Comment by trociny** [comment]
@idryomov Thanks, updated.

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
No related failures:

https://pulpito.ceph.com/dis-2022-01-18_16:06:59-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2022-01-19_08:39:11-rbd-wip-dis-testing-distro-default-smithi/

**Comment by idryomov** [comment]
jenkins test signed

**Pull request merged**